source/MaterialXContrib/MaterialXNode/MaterialXNode.cpp

```diff
@@ -121,35 +121,55 @@ bool MaterialXNode::getInternalValue(const MPlug& plug, MDataHandle& dataHandle)
 
 bool MaterialXNode::setInternalValue(const MPlug& plug, const MDataHandle& dataHandle)
 {
-    auto displayError = [this](const char* error)
+    auto createAndRegisterFragment = [this]()
     {
-        MString message("Failed to create shader for ");
-        message += typeName();
-        message += " '";
-        message += name();
-        message += "': ";
-        message += error;
-        MGlobal::displayError(message);
-    };
+        try
+        {
+            if (_documentFilePath.length() == 0)
+            {
+                return;
+            }
 
-    auto loadDocument = [this]()
-    {
-        return MaterialXMaya::loadDocument(
-            _documentFilePath.asChar(), Plugin::instance().getLibrarySearchPath()
-        );
-    };
+            if (!_document)
+            {
+                _document = MaterialXMaya::loadDocument(
+                    _documentFilePath.asChar(), Plugin::instance().getLibrarySearchPath()
+                );
+            };
 
-    auto createAndRegister = [this](mx::DocumentPtr document)
-    {
-        _materialXData.reset(new MaterialXData(
-            document,
-            _elementPath.asChar(),
-            Plugin::instance().getLibrarySearchPath()
-        ));
-
-        MaterialXMaya::registerFragment(
-            _materialXData->getFragmentName(), _materialXData->getFragmentSource()
-        );
+            if (_elementPath.length() == 0)
+            {
+                // When an empty element path is passed to MaterialXData's
```

Comment:
Avoiding the automatic renderable element lookup on attribute edits. This feature is only used by the creation command now.

Reply:
Alternative is to always create the data. Currently you have 2 references to a document if you do create the data -- one in the data and one on the node.

Reply:
Yes, but if we always create a MaterialXData then we'd lose the clean RAII for the resources that it manages. The two pointers to the document are shared pointers so I think it's small price to pay for the benefits. I guess there could be yet another object constructed for a valid document and owning MaterialXData if the element/shader is valid but that would complicate things.

Reply:
Yet another alternative would be to remove the document pointer from MaterialXData, if we could preprocess more on its creation so that that its clients don't need to know about the document to iterate samplers etc. I guess this cleanup can wait till after the demo.

Reply:
I was just pointing out that you have 2 references which could point to different things by accident, so that when clearing ownership of this you need to de-reference both cleanly. I didn't look at the logic that closely but if you feel it's safe, then it would be good to leave a note that they should be the same document.

Reply:
Agreed, I'll add a comment about this. Yes, I believe it's pretty safe because lifetime management of both the document and the MaterialXData is limited to MaterialXNode::setInternalValue() and MaterialXNode::setData().